File: backend/src/main/kotlin/com/compahunt/model/Vacancy.kt
package com.compahunt.model

import com.fasterxml.jackson.annotation.JsonIgnore
import jakarta.persistence.*
import org.hibernate.annotations.SQLRestriction
import java.math.BigDecimal
import java.time.Instant

@Entity
@Table(name = "vacancies")
@SQLRestriction("deleted = false")
data class Vacancy(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long = 0,

    @Column(nullable = false)
    val title: String,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "company_id", nullable = false)
    val company: Company,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    val user: User,

    @Column(nullable = false)
    val location: String,

    val jobType: String? = null,

    val experienceLevel: String? = null,

    @Column(columnDefinition = "TEXT")
    val description: String,

    @Column(name = "html_description", columnDefinition = "TEXT")
    val htmlDescription: String? = null,

    @ElementCollection
    @CollectionTable(name = "vacancy_requirements", joinColumns = [JoinColumn(name = "vacancy_id")])
    @Column(name = "requirement", columnDefinition = "TEXT")
    val requirements: List<String> = listOf(),

    @ElementCollection
    @CollectionTable(name = "vacancy_skills", joinColumns = [JoinColumn(name = "vacancy_id")])
    @Column(name = "skill")
    val skills: List<String> = listOf(),

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    val status: VacancyStatus = VacancyStatus.APPLIED,

    val appliedAt: Instant = Instant.now(),

    val postedDate: String? = null,

    val applicantCount: Int? = null,

    @Column(nullable = false, length = 1024)
    val url: String,

    @Embedded
    @AttributeOverrides(AttributeOverride(name = "location", column = Column(name = "salary_location")),)
    val salary: Salary? = null,

    val remoteness: String? = null,

    val industry: String? = null,

    val benefits: String? = null,

    val experience: String? = null,

    @OneToMany(mappedBy = "vacancy", cascade = [CascadeType.ALL], fetch = FetchType.LAZY)
    @JsonIgnore
    val interviews: List<Interview> = listOf(),

    @OneToMany(mappedBy = "vacancy", cascade = [CascadeType.ALL], fetch = FetchType.LAZY)
    @JsonIgnore
    val notes: List<VacancyNote> = listOf(),

    @Column(name = "manual", nullable = false, columnDefinition = "boolean default true")
    val manual: Boolean = false,

    @Column(nullable = false)
    val deleted: Boolean = false,

    val createdAt: Instant = Instant.now(),

    val updatedAt: Instant = Instant.now()
)

@Embeddable
data class Salary(
    val range: String,
    val min: BigDecimal?,
    val max: BigDecimal?,
    val currency: String?,
    val period: String?,
    val location: String,
    val monthMin: BigDecimal? = null,
    val monthMax: BigDecimal? = null
)

enum class Remoteness {
    ON_SITE,
    REMOTE,
    HYBRID
}

enum class VacancyStatus {
    APPLIED,
    VIEWED,
    PHONE_SCREEN,
    INTERVIEW,
    OFFER,
    REJECTED,
    ARCHIVED
}

File: frontend/src/types/vacancy.ts
export interface Company {
  id: string;
  name: string;
  website?: string;
  logoUrl?: string;
}

export interface Vacancy {
  id: string;
  title: string;
  company: Company;
  location?: string;
  salary?: number;
  status: VacancyStatus;
  appliedAt: string;
  description?: string;
  htmlDescription?: string;
  requirements?: string;
  benefits?: string;
  remoteness?: 'Remote' | 'On-site' | 'Hybrid';
  experience?: string;
  url?: string;
  lastUpdated?: string;
  manual?: boolean;
  sortOrder?: number;
}

export enum VacancyStatus {
  WISHLIST = 'WISHLIST',
  APPLIED = 'APPLIED',
  PHONE_SCREEN = 'PHONE_SCREEN',
  INTERVIEW = 'INTERVIEW',
  OFFER = 'OFFER',
  REJECTED = 'REJECTED',
  ARCHIVED = 'ARCHIVED'
}

export type ViewMode = 'table' | 'cards';

export interface Interview {
  id: string;
  vacancyId: string;
  vacancy?: Vacancy;
  scheduledAt: string;
  type: InterviewType;
  status: InterviewStatus;
  notes?: string;
  feedback?: string;
  duration?: number;
  meetingLink?: string;
  location?: string;
  interviewerName?: string;
  interviewerEmail?: string;
  createdAt: string;
  updatedAt: string;
}

export enum InterviewType {
  PHONE_SCREEN = 'PHONE_SCREEN',
  VIDEO_CALL = 'VIDEO_CALL',
  ON_SITE = 'ON_SITE',
  TECHNICAL = 'TECHNICAL',
  BEHAVIORAL = 'BEHAVIORAL',
  FINAL_ROUND = 'FINAL_ROUND',
  HR_INTERVIEW = 'HR_INTERVIEW'
}

export enum InterviewStatus {
  SCHEDULED = 'SCHEDULED',
  COMPLETED = 'COMPLETED',
  CANCELLED = 'CANCELLED',
  RESCHEDULED = 'RESCHEDULED',
  NO_SHOW = 'NO_SHOW'
}

export interface VacancyNote {
  id: string;
  vacancyId: string;
  content: string;
  type: NoteType;
  priority: NotePriority;
  tags?: string;
  isPrivate: boolean;
  createdAt: string;
  updatedAt: string;
}

export enum NoteType {
  GENERAL = 'GENERAL',
  OFFER_RECEIVED = 'OFFER_RECEIVED',
  REJECTION_RECEIVED = 'REJECTION_RECEIVED',
  INTERVIEW_FEEDBACK = 'INTERVIEW_FEEDBACK',
  FOLLOW_UP = 'FOLLOW_UP',
  RESEARCH = 'RESEARCH',
  SALARY_NEGOTIATION = 'SALARY_NEGOTIATION',
  CONTACT_INFO = 'CONTACT_INFO',
  COMPANY_CULTURE = 'COMPANY_CULTURE',
  NEXT_STEPS = 'NEXT_STEPS'
}

export enum NotePriority {
  LOW = 'LOW',
  MEDIUM = 'MEDIUM',
  HIGH = 'HIGH',
  URGENT = 'URGENT'
}